### ISSUE TITLE
Add ability to bound `SubstitutionTable`

### ISSUE THREAD

**Issue by Swatinem** [gimli-rs/cpp_demangle#277]
Profiling the case in https://github.com/getsentry/symbolic/issues/477, we spend 14% of the total runtime in *freeing* the `SubstitutionTable`.

We should definitely put an upper bound on how large the `SubstitutionTable` can get before we just give up and error out. As the linked issue mentions, a malicious symbol could potentially be used as an attack vector to provoke OOMs in demangling.

**Comment by khuey** [comment]
Are the vecs just absolutely enormous?

**Comment by Swatinem** [comment]
Yes they are. With default options, the substitution table easily grows above 200k entries.

We are running with a slightly increased recursion limit, and while debugging at one point I was past 400k entries.

The symbol from the original issue is this:

```
"_ZUlzjjlZZL1zStUlSt7j_Z3kjIIjIjL1vfIIEEEjzjjfjzSt7j_Z3kjIIjfjzL4t3kjIIjfjtUlSt7j_Z3kjIIjIjL1vfIIEEEjzjjfjzSt7j_Z3kjIIjfjzL4t3kjIIjfjzL4t7IjIjjzjjzSt7j_Z3kjIIjfjzStfjzSt7j_ZA3kjIIjIjL1vfIIEEEjzjjfjzSt7j_Z3kjIIjIjL1vfIIEEEjzjjfjzSt7j_Z3kjIIjfjzL4t3kjIIjzL4t7IjIjjzjjzSt7j_Z3kjIIjfjzStfjzSt7j_ZA3kjIIjIjL1vfIIEEEjzjjfjzSt7j_Z3kjIIjIjL1vfIIEEEjzjjfjzSt7j_Z3kjIIjfjzL4t3kjIIjfjzL4t7IjIjL1vfIIEEEjzjjSI"
```

It looks like this symbol is running into some mutual recursion between `Name` -> `TemplateArg` -> `TypeHandle` -> `ClassEnumType` -> `Name` which starts here:

https://github.com/gimli-rs/cpp_demangle/blob/a6213a06787bb9f6361f5c60cf7d8bd70f0f9262/src/ast.rs#L1703-L1708

The code in `Name` pushes a new substitution entry eagerly, which is never cleaned up when the call to `TemplateArgs` fails.

`TypeHandle` seems to be a bit smarter and only insert into the substitution table when its child successfully parses.

**Comment by Swatinem** [comment]
3 mil entries and counting, there is no end to it… Considering this looks like a mutual recursion, it is pretty much bounded by the recursion limit.

Its bad enough that the recursive descent parser is, well, recursing to try all the possibilities before it fails. But it is indeed the substitutions that are killing us here.

Reading through the code, I wonder if it is strictly necessary to allocate the backreference before recursing into `TemplateArgs`? As the `handle` is unused if it fails, I doubt there will be anything else referencing that substitution entry…

**Comment by Swatinem** [comment]
Inserting the substitution *after* recursing into the subtree does not work, but I guess that was overly optimistic.

I am however very surprised that `test_libiberty_demangle_187_` fails if I just `truncate` the substitutions in the error path.
So that specific symbol expect a failed `Name` parse to still put stuff into the substitution table as a side-effect? That sounds very bogus to me…
That specific symbol is:
```
_ZNK7strings8internal8SplitterINS_9delimiter5AnyOfENS_9SkipEmptyEEcvT_ISt6vectorI12basic_stringIcSt11char_traitsIcESaIcEESaISD_EEvEEv
```

**Comment by khuey** [comment]
> if I just `truncate` the substitutions in the error path

How, precisely, did you do this?

```diff
diff --git a/src/ast.rs b/src/ast.rs
index 54bd0f0..934301f 100644
--- a/src/ast.rs
+++ b/src/ast.rs
@@ -1705,7 +1705,13 @@ impl Parse for Name {
                 let idx = subs.insert(Substitutable::UnscopedTemplateName(name));
                 let handle = UnscopedTemplateNameHandle::BackReference(idx);

-                let (args, tail) = TemplateArgs::parse(ctx, subs, tail)?;
+                let (args, tail) = match TemplateArgs::parse(ctx, subs, tail) {
+                    Ok(v) => v,
+                    Err(e) => {
+                        subs.pop();
+                        return Err(e);
+                    }
+                };
                 return Ok((Name::UnscopedTemplate(handle, args), tail));
             } else {
                 return Ok((Name::Unscoped(name), tail));
```

Passes tests for me.

**Comment by Swatinem** [comment]
I was exposing/using a `truncate` call to reset the substitutions back to `idx`.
I thought that in the case of unbalanced push/pop, a `truncate` might be more bulletproof, but I guess not.

**Comment by khuey** [comment]
I think this is fixed by #284
